File: Sorting And Trees/Heaps, Priority Queues, Heap Sort/priority_queue.cpp
// Priority Queue - Min Heap implementation
// Author - rudrajit1729

// Class for representing priority queue

#include<bits/stdc++.h>
#define NONE INT_MIN
#define NINF -99999999
using namespace std;

// Class for representing priority queue
class Heap
{
	vector<int> heap;
public:
	int parent(int i) {return (i-1)/2;}
	void insertKey(int k);
	void decreaseKey(int i, int v);
	void minHeapify(int i);
	int extractMin();
	void deleteKey(int i);
	int getMin() {return heap[0];}
	void swap(int &a, int &b);
};

// Function to swap values
void Heap::swap(int &a, int &b)
{
	int t = a;
	a = b;
	b= t;
}

// Insert - Inserts key in heap
void Heap::insertKey(int k)
{
	heap.push_back(k);
	int i = heap.size() - 1;
	// Fix min heap property if violated
	while(i!=0 and heap[parent(i)] > heap[i])
		swap(heap[i], heap[parent(i)]);
}

// Decrease Key - Decrease value of key at index i to new val
void Heap::decreaseKey(int i, int v)
{
	heap[i] = v;
	// Fix min heap property if violated
	while(i!=0 and heap[parent(i)] > heap[i])
		swap(heap[i], heap[parent(i)]);
}

// Min Heapify - Function to preserve min heap structure
void Heap::minHeapify(int i)
{
	int smallest = i, left = 2 * i + 1, right = 2 * i + 1;
	int size = heap.size();

	if (left < size && heap[left] < heap[smallest])
		smallest = left;

	if (right < size && heap[right] < heap[smallest])
		smallest = right;

	if (smallest != i)
	{
		swap(heap[smallest], heap[i]);
		minHeapify(smallest);
	}
}

// Extract Min - Removes and returns minimum element from heap
int Heap::extractMin()
{
	if (heap.size() == 0) return NONE;

	int root = heap[0]; // Store root

	// Replace root with maximum element
	int maxChild = heap[heap.size() - 1];
	heap[0] = maxChild;

	// Reduce heap size and heapify
	heap.pop_back();
	minHeapify(0);

	return root;
}

// Delete Key - Deletes key at index i
void Heap::deleteKey(int i)
{
	decreaseKey(i, NINF);
	int x = extractMin();
}

int main()
{
	Heap h;
	h.insertKey(3); 
    h.insertKey(2); 
    h.deleteKey(1); 
    h.insertKey(15); 
    h.insertKey(5); 
    h.insertKey(4); 
    h.insertKey(45); 
    cout << h.extractMin() << " "; 
    cout << h.getMin() << " "; 
    h.decreaseKey(2, 1); 
    cout << h.getMin(); 
	cout<<"\nTest Passed!\n";
    return 0; 
}

File: Sorting And Trees/Heaps, Priority Queues, Heap Sort/priority_queue.py
# Priority Queue - Min Heap implementation
# Author - rudrajit1729

# Python has inbuilt library for these operations
from heapq import heappush, heappop, heapify

'''
heappop - pop and return smallest element from heap
heappush - push the value item onto the heap, maintaining heap invarient
heapify - transform list into heap, in place - linear time
'''

# Class for representing priority queue
class MinHeap:
	def __init__(self):
		self.heap = []

	# Parent of node i
	def parent(self, i):
		return (i-1)//2

	# Insert
	def insertKey(self, k):
		heappush(self.heap, k)

	# Decrease key - Decrease value of key at index i to new val
	def decreaseKey(self, i, v):
		self.heap[i] = v
		while(i!=0 and self.heap[self.parent(i)] > self.heap[i]):
			# Swap heap[i] with its parent
			self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]

	# Extract min - removes minimum element from heap and returns to caller function
	def extractMin(self):
		return heappop(self.heap)

	# Delete key - Deletes key at index i 
	def deleteKey(self, i):
		self.decreaseKey(i, float("-inf"))
		self.extractMin()

	# Get min - returns the minimum element in heap
	def getMin(self):
		return self.heap[0]

def main():
	heap = MinHeap()
	heap.insertKey(3)
	heap.insertKey(2)
	heap.deleteKey(1)
	heap.insertKey(15)
	heap.insertKey(5)
	heap.insertKey(4)
	heap.insertKey(45)

	print(heap.extractMin())
	print(heap.getMin())
	heap.decreaseKey(2, 1)
	print(heap.getMin())

if __name__ == '__main__':
	main()





